jax/_src/custom_derivatives.py

```diff
@@ -242,6 +242,7 @@ def _flatten_jvp(in_tree, *args):
     msg = ("Custom JVP rule must produce primal and tangent outputs with equal "
            "container (pytree) structures, but got {} and {} respectively.")
     raise TypeError(msg.format(out_tree, out_tree2))
+  # TODO(mattjj): compare primals' tangent types to tangent objects' types
```

Comment:
This is an unrelated bug I happened to notice. It's baked in the assumption that primal type equals tangent type, i.e. that `T a == a`, which is probably true for all inputs to custom_jvp/vjp functions given existing JAX datatypes but won't always be true.